[KEP-2400] kubectl top: add a --show-swap option

iholder101 opened the pull request:
#### What type of PR is this?
/kind feature

#### Background
A while ago, swap stats were added to summary and metric APIs as part of https://github.com/kubernetes/kubernetes/pull/118865. Although these stats are extremely valuable, they're mostly designed to be consumed by programs (like HPAs or metrics solutions), not by humans. This raised the request for a better way of understanding which nodes have swap, the swap capacity and usage, which pods consume swap space and how much, etc.

I've tried to address this need in two past attempts.
In the first one, https://github.com/kubernetes/kubernetes/pull/123963, I've tried to add swap as another `ReasourceName`, hence to display it as part of `kubectl describe node`'s status. This attempt was eventually denied with the main concern being that adding swap as a `ResourceName` is confusing since it cannot be directly/explicitly consumed (since the swap limit is auto-calculated).

In the second one, https://github.com/kubernetes/kubernetes/pull/125278, I've tried to add a swap condition to nodes. This PR was also denied with the main concern being that a condition is not suited to report such data, and that a field is recommended in this case, although it wasn't easy thinking of a good field that will benefit the API.

Eventually we agreed to defer this to the future. However, while attempting to GA the swap feature there was a hard push-back against GA without addressing these needs.

Therefore, this is my third attempt to address it. My hope is that it will suffice the need of debugability and will pave the way for the upcoming GA.

#### What this PR does / why we need it:
This PR adds adds the `--show-swap` option to the `kubectl top` subcommands.

Behind the scenes, this purely relies on consuming the stats added in https://github.com/kubernetes/kubernetes/pull/118865, making them human-readable.

Example output:
```
    > kubectl top node --show-swap
    NAME     CPU(cores)   CPU(%)   MEMORY(bytes)   MEMORY(%)   SWAP(bytes)   SWAP(%)   SWAP CAPACITY(bytes)
    node01   500m         8%       2836Mi          60%         0Mi           0%        0Mi
    node02   260m         5%       2206Mi          47%         102Mi         10%       1023Mi
```
```
    > kubectl top pod -n kube-system --show-swap
    NAME                                      CPU(cores)   MEMORY(bytes)   SWAP(bytes)
    coredns-58d5bc5cdb-5nbk4                  2m           19Mi            0Mi
    coredns-58d5bc5cdb-jsh26                  3m           37Mi            0Mi
    etcd-node01                               51m          143Mi           0Mi
    kube-apiserver-node01                     98m          824Mi           0Mi
    kube-controller-manager-node01            20m          135Mi           0Mi
    kube-proxy-ffgs2                          1m           24Mi            0Mi
    kube-proxy-fhvwx                          1m           39Mi            0Mi
    kube-scheduler-node01                     13m          69Mi            0Mi
    metrics-server-8598789fdb-d2kcj           5m           26Mi            0Mi
```

#### Which issue(s) this PR fixes:
Fixes #

#### Special notes for your reviewer:
I recommend to review per-commit.

#### Does this PR introduce a user-facing change?
```release-note
Added a `--show-swap` option to `kubectl top` subcommands
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2400-node-swap/README.md
```

iholder101 committed:
top, refactor: turn package-exposed variables to unexpose struct fields

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 committed:
kubectl top pod: add the --show-swap option

Example output:
> kubectl top pod -n kube-system --show-swap
NAME                                      CPU(cores)   MEMORY(bytes)   SWAP(bytes)
coredns-58d5bc5cdb-5nbk4                  2m           19Mi            0Mi
coredns-58d5bc5cdb-jsh26                  3m           37Mi            0Mi
etcd-node01                               51m          143Mi           0Mi
kube-apiserver-node01                     98m          824Mi           0Mi
kube-controller-manager-node01            20m          135Mi           0Mi
kube-proxy-ffgs2                          1m           24Mi            0Mi
kube-proxy-fhvwx                          1m           39Mi            0Mi
kube-scheduler-node01                     13m          69Mi            0Mi
metrics-server-8598789fdb-d2kcj           5m           26Mi            0Mi

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 committed:
kubectl top pod --show-swap: Add unit tests

Signed-off-by: Itamar Holder <iholder@redhat.com>

sftim commented:
Changelog suggestion
```diff
-Add a "--show-swap" option to kubectl top subcommands
+Added a `--show-swap` option to `kubectl top` subcommands
```

iholder101 commented on staging/src/k8s.io/kubectl/pkg/cmd/top/top_node.go:
```diff
@@ -227,3 +241,25 @@ func getNodeMetricsFromMetricsAPI(metricsClient metricsclientset.Interface, reso
+
+func getNodeSwapCapacity(nodeClient corev1client.CoreV1Interface, nodeName string) (*resource.Quantity, error) {
+	summaryBytes, err := nodeClient.RESTClient().Get().
+		AbsPath("/api/v1/nodes", nodeName, "proxy/stats/summary").
```
Interesting, thanks for the input! I share your concerns.
I'll try to figure out if there's a way of querying the metrics server directly

sftim replied:
I think we might want to change either the emitted metrics or the metrics-server, in order to make this feasible.

If we imply that users should have access to proxy to nodes, people will do it and that may not be a secure choice.

iholder101 replied:
> I think we might want to change either the emitted metrics or the metrics-server, in order to make this feasible.

I'm trying to understand what you exactly mean by that.
Is adding such a metric to [resource_metrics.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/metrics/collectors/resource_metrics.go) makes sense? I'm asking since all of the current metrics are around usage, not capacity/availability. If it is, I can add a `node_swap_available_bytes` metric which should be fairly easy I think. Then capacity can be deduced by `node_swap_available_bytes + node_swap_usage_bytes`.

Does this make sense?

sftim replied:
@iholder101 if you have capacity, ask SIG Security to take a look at this PR and comment formally.

Depending on the outcome, you may have nothing to do, or you might end up joining a conversation with various folks.

I have a concern here, but I won't have the capacity to guide all the work to address it.

iholder101 replied:
Thanks for your review! I'll make sure to do so :+1:

iholder101 replied:
Hey @sftim!
Just to update, in https://github.com/kubernetes/kubernetes/pull/129954 I'm trying to add the swap capacity to the node's status. If it will be accepted, it will eliminate the need of using Summary API.

iholder101 replied:
Hey @sftim!

I'm happy to update that https://github.com/kubernetes/kubernetes/pull/129954 is now merged. I've reshaped the PR to not call the summary API anymore, since now all of the info we need is part of the resource metrics data.

Can you PTAL?
If this can go in before code-freeze it would be ideal. Thanks!

sftim replied:
/uncc

I'm not the right person to provide LGTM here.

aojea commented:
LGTM

iholder101 committed:
kubectl top node: add the --show-swap option

Example output:
> kubectl top node --show-swap
NAME     CPU(cores)   CPU(%)   MEMORY(bytes)   MEMORY(%)   SWAP(bytes)   SWAP(%)
node01   500m         8%       2836Mi          60%         0Mi           0%
node02   260m         5%       2206Mi          47%         512Mi         50%

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 committed:
kubectl top node --show-swap: add unit tests

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 commented:
@haircommander @dchen1107 @yujuhong 

can you please have a look?

soltysh commented:
> can you please have a look?

Next time when touching kubectl area please make sure to reach out to sig-cli for general awareness and to get input sooner than just hours before code freeze.

iholder101 commented:
> > can you please have a look?
> 
> Next time when touching kubectl area please make sure to reach out to sig-cli for general awareness and to get input sooner than just hours before code freeze.

Thanks for the clarification. I wasn't aware this code is owned by sig/cli.
FWIW this is not a very high priority from my perspective, although it could be great to get it in before code freeze.

Will definitely reach out to sig-cli next time.

soltysh commented:
/triage accepted
/priority important-longterm
/label tide/merge-method-squash

soltysh commented on staging/src/k8s.io/kubectl/pkg/metricsutil/metrics_printer.go:
```diff
@@ -45,11 +40,31 @@ type ResourceMetricsInfo struct {
+	if showSwap {
+		printer.measuredResources = append(printer.measuredResources, "swap")
+		printer.nodeColumns = append(printer.nodeColumns, "SWAP(bytes)")
+		printer.nodeColumns = append(printer.nodeColumns, "SWAP(%)")
```
These two lines could easily be:
```golang
printer.nodeColumns = append(printer.nodeColumns, "SWAP(bytes)", "SWAP(%)")
```

soltysh commented on staging/src/k8s.io/kubectl/pkg/metricsutil/metrics_printer.go:
```diff
@@ -45,11 +40,31 @@ type ResourceMetricsInfo struct {
+	}
+
+	if showSwap {
+		printer.measuredResources = append(printer.measuredResources, "swap")
```
Can you make `swap` a constant, and use that instead of direct string.

iholder101 replied:
done

soltysh commented on staging/src/k8s.io/kubectl/pkg/metricsutil/metrics_printer.go:
```diff
@@ -59,25 +74,38 @@ func (printer *TopCmdPrinter) PrintNodeMetrics(metrics []metricsapi.NodeMetrics,
+	nodeColumns := printer.nodeColumns
+	if slices.Contains(measuredResources, "swap") {
+		nodeColumns = append(nodeColumns, "SWAP CAPACITY(bytes)")
+		measuredResources = append(measuredResources, "swap capacity")
```
Why you're adding `swap capacity` do `measuredResources`. You already have `swap` there, so what's it for? It only complicates your checks further in the code where you need to check for both `swap` and `swap capacity` where only `swap` (which should be a constant - see my other comment) is sufficient.

soltysh replied:
Currently if I run with what you have here I'm getting:
```
$ kubectl top node --show-swap 
NAME                 CPU(cores)   CPU(%)   MEMORY(bytes)   MEMORY(%)   SWAP(bytes)   SWAP(%)   SWAP CAPACITY(bytes)   
kind-control-plane   86m          0%       776Mi           1%          0Mi           0%        0Mi    
```
notice you have 3(!) colums for swap, where based on my understanding `SWAP(bytes)` and `SWAP CAPACITY(bytes)` will show the exact same value, because [node has only a single value for swap](https://github.com/kubernetes/kubernetes/blob/6b8e5a9457a015e5568d7ae21b94d8a4eeb41bb3/staging/src/k8s.io/api/core/v1/types.go#L6230-L6238). So that last columns is unnecessary.

soltysh commented on staging/src/k8s.io/kubectl/pkg/metricsutil/metrics_printer.go:
```diff
@@ -59,25 +74,38 @@ func (printer *TopCmdPrinter) PrintNodeMetrics(metrics []metricsapi.NodeMetrics,
-		})
+		}
+
+		if slices.Contains(measuredResources, "swap capacity") {
```
Just `swap`?

iholder101 replied:
done

soltysh commented on staging/src/k8s.io/kubectl/pkg/metricsutil/metrics_printer.go:
```diff
@@ -210,14 +241,14 @@ func printSingleResourceUsage(out io.Writer, resourceType v1.ResourceName, quant
 	case v1.ResourceCPU:
 		fmt.Fprintf(out, "%vm", quantity.MilliValue())
-	case v1.ResourceMemory:
+	case v1.ResourceMemory, "swap", "swap capacity":
```
Just `swap`?

iholder101 replied:
done

iholder101 commented:
> > Thanks for the clarification. I wasn't aware this code is owned by sig/cli.
> > FWIW this is not a very high priority from my perspective, although it could be great to get it in before code freeze.
> 
> If you want that before the freeze, once you get it fixed please ping me on slack after applying the suggested changes.

Thanks a lot for your quick review. Appreciate it a lot.
I believe your comments are now addressed.

soltysh reviewed (approved):
A few non-blocking comments for future. Currently you're always reporting 0 no matter if swap data is available or not, which might be misleading. I'm pretty sure we'll soon see a bug opened complaining about that fact. Similarly `--sort-by` currently supports only mem and cpu, and I might envision a request to add swap there too. 
That's something to consider for future to work on.

/approve

I'll let the sig-node folks have the final saying

iholder101 commented on staging/src/k8s.io/kubectl/pkg/metricsutil/metrics_printer.go:
```diff
@@ -28,28 +28,43 @@ import (
 	PodColumn       = "POD"
 )
 
+const swapMeasuredResource = "swap"
```
done!

kannon92 commented:
Thank you for the review @soltysh!

> Currently you're always reporting 0 no matter if swap data is available or not, which might be misleading. I'm pretty sure we'll soon see a bug opened complaining about that fact.

Honestly I think this is a good item to address.
If the swap feature gate is off, we should probably not display swap column.
If there is no swap on the node, we should distinguish the difference between 0swap and Not available swap.

Given that this is not a blocker for swap, I think iterating on this PR would be useful.

I also am curious about testing of this for the CLI. I am not a sig-cli dev though so I don't know if we should have e2e tests for this or unit tests are sufficient.

kannon92 commented:
> Similarly --sort-by currently supports only mem and cpu, and I might envision a request to add swap there too.

This should be a follow up issue to create and we can document it as a nice-to-have feature. Maybe a good-first-issue/help wanted?

soltysh commented:
Both are fine to address as followups. The swap info has to be explicitly requested through `--show-swap` flag, so it's not that bad as if we show this by default.

kannon92 commented on staging/src/k8s.io/kubectl/pkg/cmd/top/top_node.go:
```diff
@@ -198,6 +200,13 @@ func (o TopNodeOptions) RunTopNode() error {
 		}
+
+		swapCapacity := int64(0)
+		if n.Status.NodeInfo.Swap != nil && n.Status.NodeInfo.Swap.Capacity != nil {
```
We should check for NodeInfo also.

iholder101 replied:
I don't think we should. It's not a pointer, but rather a by-value field. Am I missing anything?

soltysh replied:
Itamar is correct here.

iholder101 commented:
Thanks for the reviews. Here are some thoughts.

> A few non-blocking comments for future. Currently you're always reporting 0 no matter if swap data is available or not, which might be misleading. I'm pretty sure we'll soon see a bug opened complaining about that fact.

@soltysh I'm not sure I understand the difference between 0 swap and "swap not available". Is there a difference? Can swap even be provisioned with 0 capacity?

Another concern here is that top commands currently does not have a mechanism to report "not available", which will complicate the PR and demand another refactoring.

Are we sure that's valuable? As a user that is using `--show-swap` explicitly, I think that 0 swap capacity means exactly the same as "swap is not provisioned".

> If the swap feature gate is off, we should probably not display swap column.

@kannon92 I tend to disagree. I don't see a reason to disallow using the explicit `--show-swap` flag if the feature gate is turned off. I even think that it's beneficial, especially after turning the FG on and off, to validate that indeed workloads aren't using swap. It can also be used to inspect node's swap capacity even if it is not accessible to k8s workloads.

iholder101 committed:
Explicitly mark swap as unavailable when necessary

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 commented:
> currently you're always reporting 0 no matter if swap data is available or not, which might be misleading.

Done. It will now look like the following:
With swap:
```
NAME    CPU(cores)   CPU(%)   MEMORY(bytes)   MEMORY(%)   SWAP(bytes)   SWAP(%)   
node1   1m           10%      2Mi             10%         1Mi           0%        
node2   5m           10%      6Mi             10%         2Mi           0%        
node3   3m           10%      4Mi             10%         0Mi           0%        
```

Withous swap:
```
NAME    CPU(cores)   CPU(%)   MEMORY(bytes)   MEMORY(%)   SWAP(bytes)   SWAP(%)       
node1   1m           10%      2Mi             10%         unavailable   unavailable   
node2   5m           10%      6Mi             10%         unavailable   unavailable   
node3   3m           10%      4Mi             10%         unavailable   unavailable   
```

I prefer to leave the `top pod` as-is TBH. It could be a useful way of ensuring pods aren't using swap after switching swap on/off, as the node will set swap capacity to nil if the feature gate is dropped:
https://github.com/kubernetes/kubernetes/blob/beef784c51f466e1a021a30c2d99132a94c233b1/pkg/kubelet/nodestatus/setters.go#L359-L363

In addition, currently the pod printer relies solely on metrics, which only say that the pod is not using any swap. To gather information about nodes we'd need to add an addition list node API call.

iholder101 commented:
> > currently you're always reporting 0 no matter if swap data is available or not, which might be misleading.
> 
> @soltysh Done. It will now look like the following: With swap:
> 
> ```
> NAME    CPU(cores)   CPU(%)   MEMORY(bytes)   MEMORY(%)   SWAP(bytes)   SWAP(%)   
> node1   1m           10%      2Mi             10%         1Mi           0%        
> node2   5m           10%      6Mi             10%         2Mi           0%        
> node3   3m           10%      4Mi             10%         0Mi           0%        
> ```
> 
> Withous swap:
> 
> ```
> NAME    CPU(cores)   CPU(%)   MEMORY(bytes)   MEMORY(%)   SWAP(bytes)   SWAP(%)       
> node1   1m           10%      2Mi             10%         unavailable   unavailable   
> node2   5m           10%      6Mi             10%         unavailable   unavailable   
> node3   3m           10%      4Mi             10%         unavailable   unavailable   
> ```
> 
> I prefer to leave the `top pod` as-is TBH. It could be a useful way of ensuring pods aren't using swap after switching swap on/off, as the node will set swap capacity to nil if the feature gate is dropped:
> 
> https://github.com/kubernetes/kubernetes/blob/beef784c51f466e1a021a30c2d99132a94c233b1/pkg/kubelet/nodestatus/setters.go#L359-L363
> 
> In addition, currently the pod printer relies solely on metrics, which only say that the pod is not using any swap. To gather information about nodes we'd need to add an addition list node API call.

@soltysh WDYT? :)

soltysh commented on staging/src/k8s.io/kubectl/pkg/metricsutil/metrics_printer.go:
```diff
@@ -210,14 +245,19 @@ func printSingleResourceUsage(out io.Writer, resourceType v1.ResourceName, quant
 
-func printPodResourcesSum(out io.Writer, total v1.ResourceList, columnWidth int) {
+func printSingleMissingResource(out io.Writer) {
+	const unavailableStr = "unavailable"
```
```suggestion
	const unavailableStr = "<unkown>"
```
is what we're mostly using, in `kubectl decribe` output, for example. So let's be consistent with that.

soltysh reviewed (commented):
Two final comments, and lets land this. @iholder101 ping me on slack when you get that updated and I'll tag.

iholder101 commented:
/test pull-kubernetes-unit

soltysh reviewed (approved):
/lgtm
/approve
